File: 2015/20/main.ts
#!/usr/bin/env -S deno run --allow-read

const inputText = (await Deno.readTextFile(
  new URL("input.txt", import.meta.url),
)).trim();

const input = Number(inputText);

const factors = (n: number): Array<number> => {
  const f: Array<number> = [];
  for (let i = 1; i <= Math.sqrt(n); i++) {
    if (n % i !== 0) continue;
    f.push(i);
    const c = n / i;
    if (c !== i) f.push(c);
  }
  return f;
};

let answerOne = 0;
for (let i = 1; i < 1_000_000; i++) {
  let presents = 0;
  for (const elf of factors(i)) presents += elf * 10;
  if (presents >= input) {
    answerOne = i;
    break;
  }
}
console.log(`Answer 1: ${answerOne}`);

let answerTwo = 0;
const elves: Array<number> = [];
for (let house = 1; house < 1_000_000; house++) {
  let presents = 0;

  for (const elf of factors(house)) {
    if (elves[elf] === 50) continue;
    // if (house % elf !== 0) continue;
    presents += elf * 11;
    elves[elf] ??= 0;
    elves[elf]++;
  }
  if (presents >= input) {
    answerTwo = house;
    break;
  }
}

console.log(`Answer 2: ${answerTwo}`);

File: 2015/20/main.zig
const std = @import("std");

const inputText = @embedFile("input.txt");
const input = std.fmt.parseInt(i32, inputText[0 .. inputText.len - 1], 10) catch 0;

// var gpa = std.heap.GeneralPurposeAllocator(.{}){};
// const allocator = gpa.allocator();
const allocator = std.heap.c_allocator;

fn factors(n: u32, list: *std.ArrayList(u32)) !void {
    for (1..std.math.sqrt(n)) |i| {
        if (@mod(n, i) != 0) continue;
        try list.append(@intCast(i));
        const c = @divFloor(n, i);
        if (c != i) try list.append(@intCast(c));
    }
}

pub fn main() !void {
    var answer_one: u32 = 0;
    for (1..1_000_000) |i| {
        var presents: u32 = 0;
        var list = std.ArrayList(u32).init(allocator);
        defer list.deinit();
        try factors(@intCast(i), &list);
        for (list.items) |elf| {
            presents += elf * 10;
        }
        if (presents >= input) {
            answer_one = @intCast(i);
            break;
        }
    }

    std.debug.print("Answer 1: {d}\n", .{answer_one});

    var elves: [1_000_000]u8 = [_]u8{0} ** 1_000_000;
    var answer_two: u32 = 0;
    for (1..1_000_000) |house| {
        var presents: u32 = 0;
        var list = std.ArrayList(u32).init(allocator);
        defer list.deinit();
        try factors(@intCast(house), &list);
        for (list.items) |elf| {
            if (elves[elf] == 50) continue;
            presents += @as(u32, @intCast(elf)) * 11;
            elves[elf] += 1;
        }
        if (presents >= input) {
            answer_two = @intCast(house);
            break;
        }
    }

    std.debug.print("Answer 2: {d}\n", .{answer_two});
}
